google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java

```diff
@@ -298,6 +298,36 @@ protected SpannerRpc getSpannerRpcV1() {
     return (SpannerRpc) getRpc();
   }
 
+  /**
+   * Returns a {@link Spanner} service object. This method will create a {@link Spanner} instance on
+   * the first call and subsequent calls to this method will return the same {@link Spanner}
+   * instance. If the instance is closed, the following call to this method will create a new {@link
+   * Spanner} instance.
+   */
+  @Override
+  public Spanner getService() {
+    Spanner spanner = super.getService();
+    if (spanner.isClosed()) {
```

Comment:
I think that the new object needs to be cached.  Is there a way to make sure that `super.getService()`'s object is refreshed?  The same comment applies to `getRpc()`

Reply:
Yes, that is done by the method call `ServiceOptions.createNewService()` on L311 (and on L326 for `getRpc()`).

Reply:
Where is the object cached?

Reply:
It's cached in the super class. The `super.getService()`, which is called on first line of the method, gets the cached object from the super class. If it is closed, the `createNewService()` method (also defined in the super class `ServiceOptions`) will refresh it.

The implementation of `ServiceOptions.getService()` looks like this:

```java
  ...
  private transient ServiceT service;
  ...
  public ServiceT getService() {
    if (service == null) {
      service = serviceFactory.create((OptionsT) this);
    }
    return service;
  }
```

Reply:
Won't the super class will cache the closed service, hence after super.service is closed, each new call to `getService()` will get a new instance?

Reply:
The easiest way to show what happens is by showing the code as if it was in one class (in reality it is split between the super class and sub class, the `createNewService()` method is in the super class). This is what is going on:

```java
  ...
  private transient ServiceT service;
  ...
  @Override
  public Spanner getService() {
    Spanner spanner = super.getService();
    if (spanner.isClosed()) {
      spanner = createNewService();
    }
    return spanner;
  }

  protected ServiceT createNewService() {
    service = serviceFactory.create((OptionsT) this);
    return service;
  }
```

Reply:
I've added an additional test case to ensure that the correct caching behavior is actually happening here: https://github.com/googleapis/google-cloud-java/blob/75d68e7ee646065f0794923520c0332a41d278d9/google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerOptionsTest.java#L89

Reply:
ugh... I reviewed the code a little closer, and debugged it.  I see what you did now.  it was a bit confusing.

Perhaps change this:

```
  @SuppressWarnings("unchecked")
  public ServiceT getService() {
    if (service == null) {
      service = serviceFactory.create((OptionsT) this);
    }
    return service;
  }

  @SuppressWarnings("unchecked")
  protected ServiceT createNewService() {
    service = serviceFactory.create((OptionsT) this);
    return service;
  }
```

to 
```
  @SuppressWarnings("unchecked")
  public ServiceT getService() {
    if (shouldRefreshService()) {
      service = serviceFactory.create((OptionsT) this);
    }
    return service;
  }

  @SuppressWarnings("unchecked")
  protected boolean shouldRefreshService() {
    return service == null;
  }
```

That would make the logic a bit less spread out and readable.

Reply:
That is a cleaner solution. The subclass does need access to the cached object, so I've included the cached service object to the signature of the `shouldRefreshService(ServiceT)` method.